src/exchange/exchange.py

```diff
@@ -158,16 +167,144 @@ def add_tool_use(self, tool_use: ToolUse) -> None:
         self.add(Message(role="assistant", content=[tool_use]))
         self.add(Message(role="user", content=[tool_result]))
 
-    def add_checkpoint(self, usage: Usage) -> None:
-        self.checkpoints.append(
+    def add_checkpoints_from_usage(self, usage: Usage) -> None:
+        """
+        Add checkpoints to the exchange based on the token counts of the last two
+        groups of messages, as well as the current token total count of the exchange
+        """
+        # we know we just appended one message as the response from the LLM
+        # so we need to create two checkpoints as we know the token counts
+        # of the last two groups of messages:
+        # 1. from the last checkpoint to the most recent user message
+        # 2. the most recent assistant message
+        last_checkpoint_end_index = (
+            self.checkpoint_data.checkpoints[-1].end_index - self.checkpoint_data.message_index_offset
+            if len(self.checkpoint_data.checkpoints) > 0
+            else -1
+        )
+        new_start_index = last_checkpoint_end_index + 1
+
+        # here, our self.checkpoint_data.total_token_count is the previous total token count from the last time
+        # that we performed a request. if we subtract this value from the input_tokens from our
+        # latest response, we know how many tokens our **1** from above is.
+        first_block_token_count = usage.input_tokens - self.checkpoint_data.total_token_count
+        second_block_token_count = usage.output_tokens
+
+        if len(self.messages) - new_start_index > 1:
+            # this case may come up if the user manipulates the message queue. e.g. if the
+            # user performs the following actions:
+            # 1. adds a message
+            # 2. calls pop_last_checkpoint
+            # 3. adds another message
+            # 4. calls reply
+            self.checkpoint_data.checkpoints.append(
+                Checkpoint(
+                    start_index=new_start_index + self.checkpoint_data.message_index_offset,
```

Comment:
Maybe a comment about the -2? If i'm thinking about this correctly, we're adding in the first checkpoint (so first group of messages). The last message (corresponding to output_tokens) would be -1.

Reply:
yup, just added some more detail here. good callout on the explanation.